collector/terminal_services.go

```diff
@@ -0,0 +1,397 @@
+		return err
+	}
+	if desc, err := c.collectTSSessionCounters(ctx, ch); err != nil {
+		log.Error("failed collecting terminal services session count metrics:", desc, err)
+		return err
+	}
+
+	// only collect CollectionBrokerPerformance if host is a Connection Broker
+	if connectionBrokerEnabled {
+		if desc, err := c.collectCollectionBrokerPerformanceCounter(ctx, ch); err != nil {
+			log.Error("failed collecting Connection Broker performance metrics:", desc, err)
+			return err
+		}
+	}
+	return nil
+}
+
+type perflibTerminalServices struct {
+	ActiveSessions   float64 `perflib:"Active Sessions"`
+	InactiveSessions float64 `perflib:"Inactive Sessions"`
+	TotalSessions    float64 `perflib:"Total Sessions"`
+}
+
+func (c *TerminalServicesCollector) collectTSSessionCount(ctx *ScrapeContext, ch chan<- prometheus.Metric) (*prometheus.Desc, error) {
+	dst := make([]perflibTerminalServices, 0)
+	err := unmarshalObject(ctx.perfObjects["Terminal Services"], &dst)
+	if err != nil {
+		return nil, err
+	}
+	if len(dst) == 0 {
+		return nil, errors.New("WMI query returned empty result set")
+	}
+
+	ch <- prometheus.MustNewConstMetric(
+		c.LocalSessionCount,
+		prometheus.GaugeValue,
+		float64(dst[0].ActiveSessions),
+		"active",
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.LocalSessionCount,
+		prometheus.GaugeValue,
+		float64(dst[0].InactiveSessions),
+		"inactive",
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.LocalSessionCount,
+		prometheus.GaugeValue,
+		float64(dst[0].TotalSessions),
+		"total",
+	)
+
+	return nil, nil
+}
+
+type perflibTerminalServicesSession struct {
+	Name                  string
+	HandleCount           float64 `perflib:"Handle Count"`
+	PageFaultsPersec      float64 `perflib:"Page Faults/sec"`
+	PageFileBytes         float64 `perflib:"Page File Bytes"`
+	PageFileBytesPeak     float64 `perflib:"Page File Bytes Peak"`
+	PercentPrivilegedTime float64 `perflib:"% Privileged Time"`
+	PercentProcessorTime  float64 `perflib:"% Processor Time"`
+	PercentUserTime       float64 `perflib:"% User Time"`
+	PoolNonpagedBytes     float64 `perflib:"Pool Nonpaged Bytes"`
+	PoolPagedBytes        float64 `perflib:"Pool Paged Bytes"`
+	PrivateBytes          float64 `perflib:"Private Bytes"`
+	ThreadCount           float64 `perflib:"Thread Count"`
+	VirtualBytes          float64 `perflib:"Virtual Bytes"`
+	VirtualBytesPeak      float64 `perflib:"Virtual Bytes Peak"`
+	WorkingSet            float64 `perflib:"Working Set"`
+	WorkingSetPeak        float64 `perflib:"Working Set Peak"`
+}
+
+func (c *TerminalServicesCollector) collectTSSessionCounters(ctx *ScrapeContext, ch chan<- prometheus.Metric) (*prometheus.Desc, error) {
+	dst := make([]perflibTerminalServicesSession, 0)
+	err := unmarshalObject(ctx.perfObjects["Terminal Services Session"], &dst)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, d := range dst {
+		ch <- prometheus.MustNewConstMetric(
+			c.HandleCount,
+			prometheus.GaugeValue,
+			float64(d.HandleCount),
```

Comment:
Redundant `float64()` cast can be removed, as variables are initialized as `float64`.